enhancements/etcd/etcd-encryption-for-separate-oauth-apis.md

```diff
@@ -0,0 +1,149 @@
+3. Update `CAO` to revision and plug `encryption-config` for its operand.
+4. Create a new deployer (`statemachine.Deployer`) called `UnionRevisionLabelPodDeployer` that will manage multiple `RevisionLabelPodDeployer`.
+5. Implement `DeployedEncryptionConfigSecret() (secret *corev1.Secret, converged bool, err error)` function that:
+   - calls `DeployedEncryptionConfigSecret` for all components
+   - returns "failure" if any component has reported an error (`err !=nil`) or hasn't yet converged (`converged == false`)
+   - returns "failure" if the `secret` resource differs among components. In oder to compare the `secret`:
+     - use `encryptionconfig.FromSecret` function that returns `EncryptionConfiguration`
+     - use `reflect.DeepEqual` function to compare `EncryptionConfigurration.Resources` for example `reflect.DeepEqual(openshiftAPIServerEncryptionCfg.Resources, oauthAPIServerEncryptionCfg.Resources)`
+   - returns "success" otherwise
+6. Change `NewRevisionLabelPodDeployer` to conditionally keep the secrets in synchornization. 
+7. Update `OAS-O` to use `UnionRevisionLabelPodDeployer` passing two `RevisionLabelPodDeployer`. The first one for `openshift-apiserver` (already existing) and the second one for `oauth-apiserver`.
+   - it will report whether all instances of `oauth-apiserver` converged to the same revision
+   - make sure that `oauth-apiserver` deployer won't synchronize the secrets
+
+### User Stories [optional]
+
+#### Story 1
+
+#### Story 2
+
+### Risks and Mitigations
+
+
+## Design Details
+
+### Test Plan
+
+To validate if `CAO` is capable of managing its own `encryption-config` we are going to create the following E2E tests:
+ - scenario 1: turn on encryption
+ - scenario 2: turn encryption on and off
+ - scenario 3: turn on encryption and force key rotation
+ - scenario 4: measure migration performance
+  
+Note: the above tests will be created based on the common test library that drives the same set of tests for `OAS-O` and `KAS-O`.
+
+To validate upgrade / downgrade path for `4.n` and `4.(n-1)` we are going to create the following E2E tests:
+ - scenario 1: take `4.(n-1)` cluster, encrypt it, upgrade to `4.n`, downgrade and upgrade again
+ - scenario 2: take `4.(n-1)` cluster, encrypt it, upgrade to `4.n`, force key rotation, downgrade to `4.(n-1)`, force key rotation and upgrade again
+ 
+Note: At the moment we don't have tests like that so creating them will be significantly harder.
+
+To validate upgrade / downgrade path for future release `4.(n+1)` we are going to create the following E2E tests:
+ - scenario 1: take `4.n` cluster, encrypt it but let `OAS-O` manage `encryption-config`, upgrade to the next version in which `CAO` will take over the config then downgrade
+ - scenario 2: take `4.n` cluster, encrypt it but let `OAS-O` manage `encryption-config`, upgrade, force key rotation, downgrade, force key rotation and upgrade again
```

Comment:
a fresh `4.n` could start off without the annotation. Wdyt? I think this reduces test effort a little as this is the standard case and the upgrade case is special.

Reply:
Wondering whether a fresh 4.n cluster can be downgraded?

Reply:
> a fresh 4.n could start off without the annotation

yes, good point, we need to covert that path as well. Okay I will update `scenario 2`.